File: tests/openssl.pl
#!/usr/bin/perl

my $TMPFILE="out-openssl.tmp";

print "Checking the OpenSSL configuration\n";

my $dir = 'tests/outputs';

opendir(DIR, $dir) or die $!;

my @opensslpolicies
    = grep {
        /openssl\./         # has openssl. in name
        && -f "$dir/$_"   # and is a file
    } readdir(DIR);

foreach my $policyfile (@opensslpolicies) {
	my $policy = $policyfile;
	$policy =~ s/-[^-]+$//;

	print "Checking policy $policy\n";

	my $tmp = do {
		local $/ = undef;
		open my $fh, "<", $dir.'/'.$policyfile
			or die "could not open $file: $!";
		<$fh>;
	};

	my %skip_test = map {$_ => 1} ("EMPTY", "GOST-ONLY");

	system("openssl ciphers $tmp >$TMPFILE 2>&1") unless exists $skip_test{$policy};
	if ($? != 0) {
		print "Error in OpenSSL policy for $policy\n";

		print STDERR "openssl ciphers error:\n";
		system("cat $TMPFILE 1>&2");
		print STDERR "ciphers: $tmp\n";
		exit 1;
	}
	unlink($TMPFILE);
}

exit 0;

File: tests/openssl.py
#!/usr/bin/python3

import re
import subprocess
import sys
from pathlib import Path

print('Checking the OpenSSL configuration')

for policy_path in Path('tests', 'outputs').glob('*-opensslcnf.txt'):
    policy = policy_path.name.removesuffix('-opensslcnf.txt')
    if policy in {'EMPTY', 'GOST-ONLY'}:
        continue
    print(f'Checking policy {policy}')

    ciphers, = re.findall(r'^CipherString = (.*)$',
                          policy_path.read_text(encoding='utf-8'),
                          re.MULTILINE)

    try:
        p = subprocess.run(['openssl', 'ciphers', ciphers],  # noqa: S607
                           check=True,
                           encoding='utf-8',
                           stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError as e:
        print(f'openssl ciphers error {e.returncode}:', file=sys.stderr)
        print(e.stdout, file=sys.stderr)
        print(f'ciphers: {ciphers}', file=sys.stderr)
        sys.exit(1)
